typed/merge.go

```diff
@@ -17,11 +17,62 @@ limitations under the License.
 package typed
 
 import (
+	"sort"
+
 	"sigs.k8s.io/structured-merge-diff/fieldpath"
 	"sigs.k8s.io/structured-merge-diff/schema"
 	"sigs.k8s.io/structured-merge-diff/value"
 )
 
+type pathElementValueMap struct {
+	members sortedPathElementValues
+}
+
+type pathElementValue struct {
+	PathElement fieldpath.PathElement
+	Value       value.Value
```

Comment:
I was going to say, it might be worth testing whether *value.Value is better for perf (less data to copy around), but below I found an even better reason.

Reply:
An alternative is to never move objects in the slice but instead keep a parallel []int containing the sorted order.

Reply:
Let's not over-optimize. Values are going to change significantly anyway.